### PR TITLE
add branches on staging page

### PR DESCRIPTION
**Pull request by dannyroberts**
at /hq/admin/system/autostaging/

### CODE REVIEW

File: corehq/apps/hqadmin/views.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 import json
 import socket
 from datetime import timedelta, date
-from collections import defaultdict
+from collections import defaultdict, namedtuple
```
Comment (dimagimon) [inline corehq/apps/hqadmin/views.py]: F401 'namedtuple' imported but unused

File: corehq/apps/hqadmin/views.py [hunk 2]
```diff
@@ -726,3 +726,29 @@ def _malt_csv_response(month, year):
...
+
+@require_superuser
+def branches_on_staging(request, template='hqadmin/branches_on_staging.html'):
+    branches = _get_branches_merged_into_autostaging()
+    return render(request, template, {'branches': branches})
+
+
+def _get_branches_merged_into_autostaging():
+    import sh
+    from collections import namedtuple
```
Comment (dimagimon) [inline corehq/apps/hqadmin/views.py]: F811 redefinition of unused 'namedtuple' from line 5

File: corehq/apps/hqadmin/templates/hqadmin/branches_on_staging.html [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<h1>Branches Currently Deployed</h1>
+<p>This page only means something on staging.</p>
+<ul>
+{% for branch in branches %}
+    <li>
+        <a href="https://github.com/dimagi/commcare-hq/compare/{{ branch.branch }}" target="_blank">{{ branch.branch }}</a>
+        (<a href="https://github.com/dimagi/commcare-hq/commit/{{ branch.commit }}">{{ branch.commit }}</a>)
+    </li>
+{% endfor %}
+</ul>
```
Comment (esoergel) [inline corehq/apps/hqadmin/templates/hqadmin/branches_on_staging.html]: This is a standalone html page, right?  Not added as an `{% include %}` or anything?  Could throw in some `<html><head></head><body></body></html>` tags to make that more obvious.

File: corehq/apps/hqadmin/views.py [hunk 4]
```diff
@@ -726,3 +726,28 @@ def _malt_csv_response(month, year):
...
+    branches = _get_branches_merged_into_autostaging()
+    return render(request, template, {'branches': branches})
+
+
+def _get_branches_merged_into_autostaging():
+    import sh
+    git = sh.git.bake(_tty_out=False)
+    # %p %s is parent hashes + subject of commit message, which will look like:
+    # <merge base> <merge head> Merge <stuff> into autostaging
+    pipe = git.log('origin/master...', grep='Merge .* into autostaging', format='%p %s')
```
Comment (esoergel) [inline corehq/apps/hqadmin/views.py]: What does `git log origin/master...` do?  Wouldn't you want to log `HEAD` (assuming this is meant to be viewed on staging)?

Comment (dannyroberts) [inline corehq/apps/hqadmin/views.py]: That's git revision range syntax. If you're on branch `autostaging`, `git log origin/master...` is short for `git log origin/master...autostaging`, which means give me the log of all changes between master and autostaging. Triple dot means symmetric difference, double dot means asymmetric difference. In this case a double dot would have worked just as well. See more here: https://git-scm.com/book/en/v2/Git-Tools-Revision-Selection#Commit-Ranges

**Comment by dannyroberts** [comment]
Thanks for the comments, @esoergel. Updated

File: corehq/apps/hqadmin/views.py (hunk 4 above)
Comment (esoergel) [inline corehq/apps/hqadmin/views.py]: ah, nice.  I've always used `git log mybranch --not master`

**Comment by esoergel** [comment]
(canceling travis builds)

File: corehq/apps/hqadmin/views.py (hunk 4 above)
Comment (dannyroberts) [inline corehq/apps/hqadmin/views.py]: Wild. I use `..` and `...` constantly, and I'd never though of using `--not` for that.

**Pull request merged**